chain/subscriber.go

```diff
@@ -0,0 +1,517 @@
+		persistInterval:    persistInterval,
+		walletAddress:      walletAddress,
+
+		// locked
+		lastSave: time.Now(),
+	}
+
+	// subscribe to contract id updates
+	subscriber.knownContracts, subscriber.unsubscribeFn, err = contracts.AddSubscriber(context.Background(), subscriber)
+	if err != nil {
+		return nil, err
+	}
+
+	return subscriber, nil
+}
+
+func (cs *Subscriber) Close() error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	cs.closed = true
+	cs.unsubscribeFn()
+	if cs.persistTimer != nil {
+		cs.persistTimer.Stop()
+		select {
+		case <-cs.persistTimer.C:
+		default:
+		}
+	}
+	if cs.nextUpdate != nil && cs.nextUpdate.HasUpdates() {
+		return cs.commit()
+	}
+	return nil
+}
+
+func (cs *Subscriber) NewContractID(id types.FileContractID) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	cs.knownContracts[id] = api.ContractStatePending
+}
+
+func (cs *Subscriber) ProcessChainApplyUpdate(cau chain.ApplyUpdate) error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	// check for shutdown, ideally this never happens since the subscriber is
+	// unsubscribed first and then closed
+	if cs.closed {
+		return errors.New("shutting down")
+	}
+
+	// set the tip
+	if cs.nextUpdate == nil {
+		cs.nextUpdate = NewChainUpdate(cau.State.Index)
+	} else {
+		cs.nextUpdate.Index = cau.State.Index
+	}
+
+	// process chain updates
+	cs.processChainApplyUpdateHostDB(cau)
+	cs.processChainApplyUpdateContracts(cau)
+	if err := cs.processChainApplyUpdateWallet(cau); err != nil {
+		return err
+	}
+
+	return cs.tryCommit()
+}
+
+func (cs *Subscriber) ProcessChainRevertUpdate(cru chain.RevertUpdate) error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	// check for shutdown, ideally this never happens since the subscriber is
+	// unsubscribed first and then closed
+	if cs.closed {
+		return errors.New("shutting down")
+	}
+
+	// set the tip
+	if cs.nextUpdate == nil {
+		cs.nextUpdate = NewChainUpdate(cru.State.Index)
+	} else {
+		cs.nextUpdate.Index = cru.State.Index
+	}
+
+	// process chain updates
+	cs.processChainRevertUpdateHostDB(cru)
+	cs.processChainRevertUpdateContracts(cru)
+	if err := cs.processChainRevertUpdateWallet(cru); err != nil {
+		return err
+	}
+
+	return cs.tryCommit()
+}
+
+func (cs *Subscriber) Subscribe() (func(), error) {
+	index, err := cs.cs.ChainIndex()
+	if err != nil {
+		return nil, err
+	}
+	if err := cs.sync(index); err != nil {
+		return nil, fmt.Errorf("failed to subscribe to chain manager: %w", err)
+	}
+
+	reorgChan := make(chan types.ChainIndex, 1)
+	go func() {
+		for range reorgChan {
+			lastTip, err := cs.cs.ChainIndex()
+			if err != nil {
+				cs.logger.Error("failed to get last committed index", zap.Error(err))
+				continue
+			}
+			if err := cs.sync(lastTip); err != nil {
+				cs.logger.Error("failed to sync store", zap.Error(err))
+			}
+		}
+	}()
+
+	return cs.cm.OnReorg(func(index types.ChainIndex) {
+		select {
+		case reorgChan <- index:
+		default:
+		}
+	}), nil
+}
+
+func (cs *Subscriber) TriggerSync() {
```

Comment:
I think this needs to be a `Run` or some other sort of loop that asks for new updates every 10 seconds or so. We shouldn't need to ever trigger the subscriber manually.